File: src/example/lua/websocket_server.lua
#!/usr/bin/lua

local ev = require("ev")
local evmg = require("evmongoose")
local loop = ev.Loop.default
local mgr = evmg.init()

local page = [[
<!DOCTYPE html>
<html>
<head>
	<meta charset="utf-8">
	<title>WebSocket Test</title>
	<script>
		var ws = null;
		function connect(btn){
			if (ws) {
				ws.close()
				document.getElementById("btn").value="Connect"
				ws = null;
				return;
			}
			document.getElementById("btn").value = "Disconnect";
			
			ws = new WebSocket("ws://" + location.host);
			ws.onopen = function() {
				console.log("open success");
				ws.send("Hello Evmg");
			};
			
			ws.onmessage = function(evt) {
				console.log("read:" + evt.data);
			};
			
			ws.onclose = function () {
				console.log("server close");
			};
		}
	</script>
	</head>
	<body>
		<input type="button" id="btn" value="Connect" onclick="connect()"/>
	</body>
</html>
]]

local op_str = {
	[evmg.WEBSOCKET_OP_CONTINUE] = "continue",
	[evmg.WEBSOCKET_OP_TEXT] = "text",
	[evmg.WEBSOCKET_OP_BINARY] = "binary",
	[evmg.WEBSOCKET_OP_CLOSE] = "close",
	[evmg.WEBSOCKET_OP_PING] = "ping",
	[evmg.WEBSOCKET_OP_PONG] = "pong",
	[-1] = "unknown"
}

local function ev_handle(nc, event, msg)
	if event == evmg.MG_EV_HTTP_REQUEST then
		mgr:send_head(nc, 200, #page)
		mgr:print(nc, page)
		return true
		
	elseif event == evmg.MG_EV_WEBSOCKET_HANDSHAKE_REQUEST then
		print("method:", msg.method)
		print("uri:", msg.uri)
		print("proto:", msg.proto)
		print("query_string:", msg.query_string)

		for k, v in pairs(msg.headers) do
			print(k, v)
		end

	elseif event == evmg.MG_EV_WEBSOCKET_CONTROL_FRAME then
		print("recv control frame:", op_str[msg.op])
	elseif event == evmg.MG_EV_WEBSOCKET_FRAME then
		print("recv frame", msg.data, op_str[msg.op])
		mgr:send_websocket_frame(nc, "I is evmg", evmg.WEBSOCKET_OP_TEXT)
	end
end

mgr:bind("8000", ev_handle, {proto = "http"})
print("Listen on http 8000...")

ev.Signal.new(function(loop, sig, revents)
	loop:unloop()
end, ev.SIGINT):start(loop)

loop:loop()

mgr:destroy()

print("exit...")

File: src/example/websocket_server.c
#include <mongoose.h>

static const char *s_http_port = "8000";
static struct mg_serve_http_opts s_http_server_opts;

static const char *index_page = "<!DOCTYPE html><html><head>"
	"<meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"/><title>Websocket Test</title></head>"
	"<body><button type=\"button\" onclick=\"connect()\">Connect</button>"
	"<script>function connect(){var ws = new WebSocket(\"ws://\" + location.host);ws.onopen = function() {"
	"console.log(\"connect ok\");ws.send(\"Hello WebSocket\");};ws.onmessage = function (evt) {"
	"console.log(\"recv:\"+evt.data);};ws.onclose = function(){console.log(\"close\");};}</script></body></html>";

static void ev_handler(struct mg_connection *nc, int ev, void *ev_data)
{	
	switch (ev) {
	case MG_EV_HTTP_REQUEST:
			mg_printf(nc, "HTTP/1.1 200 OK\r\n"
							  "Content-Length: %d\r\n\r\n"
							  "%s", (int)strlen(index_page), index_page);
			break;
			
	case MG_EV_WEBSOCKET_HANDSHAKE_DONE: {
			char buf[] = "Hello Websocket Client";
			printf("New websocket connection:%p\n", nc);
			mg_send_websocket_frame(nc, WEBSOCKET_OP_TEXT, buf, strlen(buf));
			break;
		}
	case MG_EV_WEBSOCKET_FRAME: {
			struct websocket_message *wm = (struct websocket_message *)ev_data;
			printf("recv websocket:%.*s\n", (int)wm->size, wm->data);
			break;
		}
	default:
		break;		
	}
}

static void signal_cb(struct ev_loop *loop, ev_signal *w, int revents)
{
	printf("Got signal: %d\n", w->signum);
	ev_break(loop, EVBREAK_ALL);
}

int main(int argc, char **argv)
{
	struct ev_loop *loop = EV_DEFAULT;
	ev_signal sig_watcher;
	struct mg_mgr mgr;
	struct mg_connection *nc;

	ev_signal_init(&sig_watcher, signal_cb, SIGINT);
	ev_signal_start(loop, &sig_watcher);
	
	mg_mgr_init(&mgr, NULL, loop);
	
	printf("Starting web server on port %s\n", s_http_port);
	nc = mg_bind(&mgr, s_http_port, ev_handler);
	if (nc == NULL) {
		printf("Failed to create listener\n");
		return 1;
	}

	// Set up HTTP server parameters
	mg_set_protocol_http_websocket(nc);
	s_http_server_opts.document_root = ".";  // Serve current directory
	s_http_server_opts.enable_directory_listing = "yes";

	ev_run(loop, 0);

	printf("Server exit...\n");
	
	mg_mgr_free(&mgr);
		
	return 0;
}

